crates/iceberg/src/scan.rs

```diff
@@ -197,134 +193,103 @@ impl<'a> TableScanBuilder<'a> {
-            self.filter.clone(),
-            self.case_sensitive,
-        )?;
-
-        let mut partition_filter_cache = PartitionFilterCache::new();
-        let mut manifest_evaluator_cache = ManifestEvaluatorCache::new();
-        let mut expression_evaluator_cache = ExpressionEvaluatorCache::new();
-
-        let field_ids = self.field_ids.clone();
-        let bound_predicates = self.bound_predicates.clone();
-
-        Ok(try_stream! {
-            let manifest_list = context
-                .snapshot
-                .load_manifest_list(&context.file_io, &context.table_metadata)
-                .await?;
-
-            for entry in manifest_list.entries() {
-                if !Self::content_type_is_data(entry) {
-                    continue;
-                }
-
-                let partition_spec_id = entry.partition_spec_id;
-
-                let partition_filter = partition_filter_cache.get(
-                    partition_spec_id,
-                    &context,
-                )?;
-
-                if let Some(partition_filter) = partition_filter {
-                    let manifest_evaluator = manifest_evaluator_cache.get(
-                        partition_spec_id,
-                        partition_filter,
-                    );
-
-                    if !manifest_evaluator.eval(entry)? {
-                        continue;
-                    }
-                }
+        // used to stream ManifestEntryContexts between stages of the file plan operation
+        let (manifest_entry_ctx_tx, manifest_entry_ctx_rx) =
+            channel(CONCURRENCY_LIMIT_MANIFEST_FILES);
+        // used to stream the results back to the caller
+        let (file_scan_task_tx, file_scan_task_rx) = channel(CONCURRENCY_LIMIT_MANIFEST_ENTRIES);
+
+        let manifest_list = self.plan_context.get_manifest_list().await?;
+
+        // get the [`ManifestFile`]s from the [`ManifestList`], filtering out any
+        // whose content type is not Data or whose partitions cannot match this
+        // scan's filter
+        let manifest_file_contexts = self
+            .plan_context
+            .build_manifest_file_contexts(manifest_list, manifest_entry_ctx_tx)?;
+
+        // Concurrently load all [`Manifest`]s and stream their [`ManifestEntry`]s
+        futures::stream::iter(manifest_file_contexts)
+            .try_for_each_concurrent(CONCURRENCY_LIMIT_MANIFEST_FILES, |ctx| async move {
+                ctx.fetch_manifest_and_stream_manifest_entries().await
+            })
+            .await?;
+
+        // Process the [`ManifestEntry`] stream in parallel
+        manifest_entry_ctx_rx
+            .map(|me_ctx| Ok((me_ctx, file_scan_task_tx.clone())))
+            .try_for_each_concurrent(
+                CONCURRENCY_LIMIT_MANIFEST_ENTRIES,
+                |(manifest_entry_context, tx)| async move {
+                    crate::runtime::spawn(async move {
+                        Self::process_manifest_entry(manifest_entry_context, tx).await
```

Comment:
As a user, I prefer to manage logic like this myself because I have my own pipeline with specific resource allocation and concurrency limits. So I think @liurenjie1024's concern is true. Maybe we can make those behavior optional and disable by default. Users who want iceberg-rust to do this can enable with some config or flags.

Reply:
If I make the concurrncy limits configurable in the builder as @liurenjie1024 suggested above, then we could have a `with_no_concurrency` method on the builder. Would this be acceptable?

Reply:
In terms of having non-concurrent as the default option, my preference would be for the library to be able to choose the optimally performant (probably the number of cores?) as the default rather than no concurrncy. No concurrncy feels like a more advanced, non-default choice. @liurenjie1024 what are your thoughts?

Reply:
Regarding the large number of tasks - there are a large number generated over the duration of the plan, but only a fixed small number that are in existence at one time.

Reply:
Yeah you're right, `with_no_concurrency()` is a bit pointless when you can just set `with_max_concurrency(1)`! 😂